File: hphp/hack/src/providers/file_provider.ml
(*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)

open Hh_prelude

(*
    Shared memory heap containing the contents of files.
    Acts as a sort of caching facade which is filled on-demand
    as contents are needed - The "cache" is filled by loading from
    the file system.
*)

exception File_provider_stale

module FileHeap = struct
  include
    SharedMem.Heap
      (SharedMem.ImmediateBackend (SharedMem.Evictable)) (Relative_path.S)
      (struct
        type t = string

        let description = "File"
      end)

  let replace_nonatomic key value =
    if mem key then remove key;
    add key value
end

let read_file_contents_from_disk (fn : Relative_path.t) : string option =
  try Some (Sys_utils.cat (Relative_path.to_absolute fn)) with
  | _ -> None

let get_contents ?(force_read_disk = false) fn =
  match Provider_backend.get () with
  | Provider_backend.Analysis
  | Provider_backend.Pessimised_shared_memory _
  | Provider_backend.Shared_memory ->
    let from_cache =
      if force_read_disk then
        None
      else
        FileHeap.get fn
    in
    (match from_cache with
    | Some contents -> Some contents
    | None ->
      let contents =
        Option.value (read_file_contents_from_disk fn) ~default:""
      in
      Some contents)
  | Provider_backend.Rust_provider_backend backend ->
    Some (Rust_provider_backend.File.get_contents backend fn)
  | Provider_backend.Local_memory _
  | Provider_backend.Decl_service _ ->
    read_file_contents_from_disk fn

let provide_file_for_tests fn contents =
  match Provider_backend.get () with
  | Provider_backend.Analysis -> failwith "invalid"
  | Provider_backend.Pessimised_shared_memory _
  | Provider_backend.Shared_memory ->
    FileHeap.replace_nonatomic fn contents
  | Provider_backend.Rust_provider_backend backend ->
    Rust_provider_backend.File.provide_file_for_tests backend fn contents
  | Provider_backend.Local_memory _
  | Provider_backend.Decl_service _ ->
    failwith
      "File_provider.provide_file_for_tests not supported with local/decl memory provider"

let remove_batch paths =
  match Provider_backend.get () with
  | Provider_backend.Analysis -> failwith "invalid"
  | Provider_backend.Pessimised_shared_memory _
  | Provider_backend.Shared_memory ->
    FileHeap.remove_batch paths
  | Provider_backend.Rust_provider_backend backend ->
    Rust_provider_backend.File.remove_batch backend paths
  | Provider_backend.Local_memory _
  | Provider_backend.Decl_service _ ->
    failwith
      "File_provider.remove_batch not supported with local/decl memory provider"

let local_changes_push_sharedmem_stack () = FileHeap.LocalChanges.push_stack ()

let local_changes_pop_sharedmem_stack () = FileHeap.LocalChanges.pop_stack ()

File: hphp/hack/src/rupro/hackrs/file_provider.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.

use ocamlrep_derive::{FromOcamlRep, ToOcamlRep};
use pos::RelativePath;
use std::fmt::Debug;
use std::marker::{Send, Sync};

mod provider;
pub use provider::PlainFileProvider;

#[derive(Clone, Debug, ToOcamlRep, FromOcamlRep)]
pub enum FileType {
    Disk(bstr::BString),
    Ide(bstr::BString),
}

pub type Result<T, E = Error> = std::result::Result<T, E>;

#[derive(thiserror::Error, Debug)]
pub enum Error {
    #[error("{0}")]
    IoError(#[from] std::io::Error),
}

/// Acts as a sort of caching facade which is filled on-demand as contents are
/// needed. The "cache" is (or rather, might be) filled by loading from the file
/// system if the file isn't open in the IDE (otherwise use the IDE contents).
/// That is, any IDE version of a file takes precedence over the file system's.
// note(sf, 2022-04-28): c.f. hphp/hack/src/providers/file_provider.ml
pub trait FileProvider: Debug + Send + Sync {
    /// Lookup a path.
    fn get(&self, file: RelativePath) -> Option<FileType>;

    /// If a cached entry for the provided path exists then return its contents.
    /// If not, try to read the contents from disk. Reading from disk might
    /// produce an `Error::IoError(_)` result.
    fn get_contents(&self, file: RelativePath) -> Result<bstr::BString>;

    /// Register `file` as a disk file containing `contents`.
    fn provide_file_for_tests(&self, file: RelativePath, contents: bstr::BString);

    /// Register `file` as an IDE file containing `contents`.
    fn provide_file_for_ide(&self, file: RelativePath, contents: bstr::BString);

    /// If `file_type` is an IDE file, register it. If it is a disk file, do
    /// nothing.
    fn provide_file_hint(&self, file: RelativePath, file_type: FileType);

    /// Associate each path in `files` with `None`.
    fn remove_batch(&self, files: &[RelativePath]);

    /// Save the current set of changes on the stack then clear it.
    fn push_local_changes(&self);

    /// Clear the current set of changes.
    fn pop_local_changes(&self);
}
